### ISSUE TITLE
How to return Lists with subscriptions

### ISSUE THREAD

**Issue by Patrick-Spiegel** [graphql-dotnet/graphql-dotnet#770]
## Summary

I am trying to return Lists as Subscription result. Is that even possible? 

## Relevant information

My Field Definition

            AddField(new EventStreamFieldType
            {
                Name = valueChanged_SUBSCRIPTION,
                Arguments = new QueryArguments(
                new QueryArgument<InputServiceNodeID> { Name = targets_Argument }
             ),
                Type = typeof(ListGraphType<LiveValueBase>),
                Resolver = new FuncFieldResolver<List<LiveValueBaseContainer>>(ResolveLiveValueBaseContainer),
                Subscriber = new EventStreamResolver<List<LiveValueBaseContainer>>(SubscribeLiveValueChanged)
            });

My Resolver

        private List<LiveValueBaseContainer> ResolveLiveValueBaseContainer(ResolveFieldContext arg)
        {
            var container = arg.Source as List<LiveValueBaseContainer>;
            return container;
        }

My Publisher

                _observer.OnNext(new List<LiveValueBaseContainer>
                {
                    new LiveValueBaseContainer
                    {
                        Quality = LiveValueQuality.Bad
                    }
                });

`LiveValueQuality` is an enum. `LiveValueBase` extends from `ObjectGraphType` and takes `LiveValueBaseContainer` as generic argument. the `LiveValueBaseContainer` is a simple class containing a property with this  enum. 

The value that I get as return  value is now always null. However if I remove the `List` and the `ListGraphType` I get the enum value returned correctly. This approach works in a similar way for Queries.
Did I do something wrong?

### Environment (if relevant)

Version: 2.0.0-alpha-978

**Comment by Patrick-Spiegel** [comment]
Is this not supported? A reply would be kind :)

**Comment by joemcbride** [comment]
You should be able to return anything from a subscription that you can return as a "normal" GraphQL query.  That being said, I haven't tried this exact scenario.  If you can provide a test that shows this not working that would be helpful in tracking down any issue there may be.

**Comment by BitPuffin** [comment]
Just ran into this. Can confirm it doesn't work when you return a list (or array).

**Comment by BitPuffin** [comment]
Should a new issue be created for this or should it stay here. Was trying to set up a test in a PR but struggled getting building working on linux :/

**Comment by joemcbride** [comment]
@BitPuffin We can just use this issue.  Are you using the [server project](https://github.com/graphql-dotnet/server) for subscriptions?

**Comment by BitPuffin** [comment]
@joemcbride Yep that's what we're using.

**Comment by BitPuffin** [comment]
PR with failing test incoming

**Comment by ztripez** [comment]
Did some digging. In SubscriptionExecutionStrategy line 63-85:
```

                var resolveContext = new ResolveEventStreamContext
                {
                    FieldName = node.Field.Name,
                    FieldAst = node.Field,
                    FieldDefinition = node.FieldDefinition,
                    ReturnType = node.FieldDefinition.ResolvedType,
                    ParentType = node.GraphType as IObjectGraphType,
                    Arguments = arguments,
                    Source = source,
                    Schema = context.Schema,
                    Document = context.Document,
                    Fragments = context.Fragments,
                    RootValue = context.RootValue,
                    UserContext = context.UserContext,
                    Operation = context.Operation,
                    Variables = context.Variables,
                    CancellationToken = context.CancellationToken,
                    Metrics = context.Metrics,
                    Errors = context.Errors,
                    Path = node.Path
                };

```
 `ParentType = node.GraphType as IObjectGraphType` becomes null if it's a ListGraphType due to ListGraphType don't implements IObjectGraphType. 

Changing the base class for ListGraphType to GraphObjectType make the test in PR: https://github.com/graphql-dotnet/graphql-dotnet/pull/849 to work, 

However 2 test fails after:

```
SchemaIntrospectionTests validate_core_schema
Test Name:	GraphQL.Tests.Introspection.SchemaIntrospectionTests.validate_core_schema
Test FullName:	GraphQL.Tests.Introspection.SchemaIntrospectionTests.validate_core_schema
Test Source:	D:\code\graphql-dotnet\src\GraphQL.Tests\Introspection\SchemaIntrospectionTests.cs : line 12
Test Outcome:	Failed
Test Duration:	0:00:00,053

Result StackTrace:	
at GraphQL.Tests.Introspection.SchemaIntrospectionTests.ShouldBe(String actual, String expected) in D:\code\graphql-dotnet\src\GraphQL.Tests\Introspection\SchemaIntrospectionTests.cs:line 44
   at GraphQL.Tests.Introspection.SchemaIntrospectionTests.validate_core_schema() in D:\code\graphql-dotnet\src\GraphQL.Tests\Introspection\SchemaIntrospectionTests.cs:line 23
Result Message:	
Assert.Equal() Failure
                                 ↓ (pos 4183)
Expected: ···           "kind": "LIST",\r\n                "name": null,\r\n  ···
Actual:   ···           "kind": "OBJECT",\r\n                "name": null,\r\n···
                                 ↑ (pos 4183)

Test Name:	GraphQL.Tests.StarWars.StarWarsIntrospectionTests.allows_querying_the_schema_for_object_fields
Test FullName:	GraphQL.Tests.StarWars.StarWarsIntrospectionTests.allows_querying_the_schema_for_object_fields
Test Source:	D:\code\graphql-dotnet\src\GraphQL.Tests\StarWars\StarWarsIntrospectionTests.cs : line 92
Test Outcome:	Failed
Test Duration:	0:00:00,119

Result StackTrace:	
at Shouldly.ShouldBeStringTestExtensions.ExecuteAssertion(IAssertion assertion, Func`1 customMessage)
   at GraphQL.Tests.QueryTestBase`2.AssertQuery(String query, ExecutionResult expectedExecutionResult, Inputs inputs, Object root, Object userContext, CancellationToken cancellationToken, IEnumerable`1 rules) in D:\code\graphql-dotnet\src\GraphQL.Tests\QueryTestBase.cs:line 155
   at GraphQL.Tests.QueryTestBase`2.AssertQuerySuccess(String query, String expected, Inputs inputs, Object root, Object userContext, CancellationToken cancellationToken, IEnumerable`1 rules) in D:\code\graphql-dotnet\src\GraphQL.Tests\QueryTestBase.cs:line 55
   at GraphQL.Tests.StarWars.StarWarsIntrospectionTests.allows_querying_the_schema_for_object_fields() in D:\code\graphql-dotnet\src\GraphQL.Tests\StarWars\StarWarsIntrospectionTests.cs:line 152
Result Message:	
Shouldly.ShouldAssertException : "{
  "data": {
    "__type": {
      "name": "Droid",
      "fields": [
        {
          "name": "appearsIn",
          "type": {
            "name": null,
            "kind": "OBJECT"
          }
        },
        {
          "name": "friends",
          "type": {
            "name": null,
            "kind": "OBJECT"
          }
        },
        {
          "name": "id",
          "type": {
            "name": null,
            "kind": "NON_NULL"
          }
        },
        {
          "name": "name",
          "type": {
            "name": "String",
            "kind": "SCALAR"
          }
        },
        {
          "name": "primaryFunction",
          "type": {
            "name": "String",
            "kind": "SCALAR"
          }
        }
      ]
    }
  }
}"
    should be
"{
  "data": {
    "__type": {
      "name": "Droid",
      "fields": [
        {
          "name": "appearsIn",
          "type": {
            "name": null,
            "kind": "LIST"
          }
        },
        {
          "name": "friends",
          "type": {
            "name": null,
            "kind": "LIST"
          }
        },
        {
          "name": "id",
          "type": {
            "name": null,
            "kind": "NON_NULL"
          }
        },
        {
          "name": "name",
          "type": {
            "name": "String",
            "kind": "SCALAR"
          }
        },
        {
          "name": "primaryFunction",
          "type": {
            "name": "String",
            "kind": "SCALAR"
          }
        }
      ]
    }
  }
}"
    but was not
    difference
Showing some of the 432 differences
Difference     |                 |    |    |    |    |    |    |    |    |                                            |    |        
               |                \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/                                          \|/  \|/       
Index          | ...  186  187  188  189  190  191  192  193  194  195  196  197  198  199  200  201  202  203  204  205  206  ...  
Expected Value | ...  \s   "    L    I    S    T    "    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   \s   }    \r   ...  
Actual Value   | ...  \s   "    O    B    J    E    C    T    "    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   \s   ...  
Expected Code  | ...  32   34   76   73   83   84   34   13   10   32   32   32   32   32   32   32   32   32   32   125  13   ...  
Actual Code    | ...  32   34   79   66   74   69   67   84   34   13   10   32   32   32   32   32   32   32   32   32   32   ...  

Difference     |                                     |    |    |    |    |    |                                  |    |    |        
               |                                    \|/  \|/  \|/  \|/  \|/  \|/                                \|/  \|/  \|/       
Index          | ...  210  211  212  213  214  215  216  217  218  219  220  221  222  223  224  225  226  227  228  229  230  ...  
Expected Value | ...  \s   \s   \s   \s   \s   \s   }    ,    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   {    \r   \n   ...  
Actual Value   | ...  \s   \s   \s   \s   \s   \s   \s   \s   }    ,    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   {    ...  
Expected Code  | ...  32   32   32   32   32   32   125  44   13   10   32   32   32   32   32   32   32   32   123  13   10   ...  
Actual Code    | ...  32   32   32   32   32   32   32   32   125  44   13   10   32   32   32   32   32   32   32   32   123  ...  

Difference     |       |                                            |    |    |    |    |    |    |    |    |    |    |    |        
               |      \|/                                          \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/       
Index          | ...  232  233  234  235  236  237  238  239  240  241  242  243  244  245  246  247  248  249  250  251  252  ...  
Expected Value | ...  \s   \s   \s   \s   \s   \s   \s   \s   \s   "    n    a    m    e    "    :    \s   "    f    r    i    ...  
Actual Value   | ...  \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   \s   "    n    a    m    e    "    :    \s   "    f    ...  
Expected Code  | ...  32   32   32   32   32   32   32   32   32   34   110  97   109  101  34   58   32   34   102  114  105  ...  
Actual Code    | ...  10   32   32   32   32   32   32   32   32   32   32   34   110  97   109  101  34   58   32   34   102  ...  

Difference     |       |    |    |    |    |    |    |    |    |    |    |    |    |    |    |                                      
               |      \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/                                     
Index          | ...  248  249  250  251  252  253  254  255  256  257  258  259  260  261  262  263  264  265  266  267  268  ...  
Expected Value | ...  \s   "    f    r    i    e    n    d    s    "    ,    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   ...  
Actual Value   | ...  "    :    \s   "    f    r    i    e    n    d    s    "    ,    \r   \n   \s   \s   \s   \s   \s   \s   ...  
Expected Code  | ...  32   34   102  114  105  101  110  100  115  34   44   13   10   32   32   32   32   32   32   32   32   ...  
Actual Code    | ...  34   58   32   34   102  114  105  101  110  100  115  34   44   13   10   32   32   32   32   32   32   ...  

Difference     |                      |    |    |    |    |    |    |    |    |    |    |    |    |                                 
               |                     \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/                                
Index          | ...  268  269  270  271  272  273  274  275  276  277  278  279  280  281  282  283  284  285  286  287  288  ...  
Expected Value | ...  \s   \s   \s   "    t    y    p    e    "    :    \s   {    \r   \n   \s   \s   \s   \s   \s   \s   \s   ...  
Actual Value   | ...  \s   \s   \s   \s   \s   "    t    y    p    e    "    :    \s   {    \r   \n   \s   \s   \s   \s   \s   ...  
Expected Code  | ...  32   32   32   34   116  121  112  101  34   58   32   123  13   10   32   32   32   32   32   32   32   ...  
Actual Code    | ...  32   32   32   32   32   34   116  121  112  101  34   58   32   123  13   10   32   32   32   32   32   ...  

Difference     |                 |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |                  
               |                \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/                 
Index          | ...  292  293  294  295  296  297  298  299  300  301  302  303  304  305  306  307  308  309  310  311  312  ...  
Expected Value | ...  \s   \s   "    n    a    m    e    "    :    \s   n    u    l    l    ,    \r   \n   \s   \s   \s   \s   ...  
Actual Value   | ...  \s   \s   \s   \s   "    n    a    m    e    "    :    \s   n    u    l    l    ,    \r   \n   \s   \s   ...  
Expected Code  | ...  32   32   34   110  97   109  101  34   58   32   110  117  108  108  44   13   10   32   32   32   32   ...  
Actual Code    | ...  32   32   32   32   34   110  97   109  101  34   58   32   110  117  108  108  44   13   10   32   32   ...  

Difference     |            |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |        
               |           \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/       
Index          | ...  320  321  322  323  324  325  326  327  328  329  330  331  332  333  334  335  336  337  338  339  340  ...  
Expected Value | ...  \s   "    k    i    n    d    "    :    \s   "    L    I    S    T    "    \r   \n   \s   \s   \s   \s   ...  
Actual Value   | ...  \s   \s   \s   "    k    i    n    d    "    :    \s   "    O    B    J    E    C    T    "    \r   \n   ...  
Expected Code  | ...  32   34   107  105  110  100  34   58   32   34   76   73   83   84   34   13   10   32   32   32   32   ...  
Actual Code    | ...  32   32   32   34   107  105  110  100  34   58   32   34   79   66   74   69   67   84   34   13   10   ...  

Difference     |       |    |    |         |    |    |                        |    |    |    |    |    |    |    |                  
               |      \|/  \|/  \|/       \|/  \|/  \|/                      \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/                 
Index          | ...  347  348  349  350  351  352  353  354  355  356  357  358  359  360  361  362  363  364  365  366  367  ...  
Expected Value | ...  }    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   }    ,    \r   \n   \s   \s   \s   \s   \s   \s   ...  
Actual Value   | ...  \s   \s   \s   \s   }    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   }    ,    \r   \n   \s   \s   ...  
Expected Code  | ...  125  13   10   32   32   32   32   32   32   32   32   125  44   13   10   32   32   32   32   32   32   ...  
Actual Code    | ...  32   32   32   32   125  13   10   32   32   32   32   32   32   32   32   125  44   13   10   32   32   ...  

Difference     |       |    |         |    |    |                                  |    |    |    |    |    |    |    |    |        
               |      \|/  \|/       \|/  \|/  \|/                                \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/       
Index          | ...  371  372  373  374  375  376  377  378  379  380  381  382  383  384  385  386  387  388  389  390  391  ...  
Expected Value | ...  \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   \s   "    n    a    m    e    "    :    \s   "    ...  
Actual Value   | ...  \s   \s   \s   {    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   \s   "    n    a    m    e    ...  
Expected Code  | ...  13   10   32   32   32   32   32   32   32   32   32   32   34   110  97   109  101  34   58   32   34   ...  
Actual Code    | ...  32   32   32   123  13   10   32   32   32   32   32   32   32   32   32   32   34   110  97   109  101  ...  

Difference     |       |    |    |    |    |    |    |    |    |    |    |    |                                  |    |    |        
               |      \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/  \|/                                \|/  \|/  \|/       
Index          | ...  390  391  392  393  394  395  396  397  398  399  400  401  402  403  404  405  406  407  408  409  410  ...  
Expected Value | ...  \s   "    i    d    "    ,    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   \s   "    t    y    ...  
Actual Value   | ...  m    e    "    :    \s   "    i    d    "    ,    \r   \n   \s   \s   \s   \s   \s   \s   \s   \s   \s   ...  
Expected Code  | ...  32   34   105  100  34   44   13   10   32   32   32   32   32   32   32   32   32   32   34   116  121  ...  
Actual Code    | ...  109  101  34   58   32   34   105  100  34   44   13   10   32   32   32   32   32   32   32   32   32   ...






```

**Comment by kenrim626** [comment]
Did anyone have a further look at this issue? @ztripez showed that it is there and he even provided a possible solution. I think we should discuss what kind of functionality was destroyed with it and if we can find an alternative that fixes both (the original issue and the ones arisen from the possible solution).

**Comment by Jonatthu** [comment]
@joemcbride  @sungam3r  I am still having this issue in version 2.x.x, Is this something that has been fixed in version 3?

**Comment by Shane32** [comment]
Yes this was fixed in PR #989
